diff --git a/++dfb/idirectfbfont.cpp b/++dfb/idirectfbfont.cpp
index 06f67ec..5367307 100644
--- a/++dfb/idirectfbfont.cpp
+++ b/++dfb/idirectfbfont.cpp
@@ -52,6 +52,12 @@ int IDirectFBFont::GetHeight() const
      return height;
 }
 
+void IDirectFBFont::GetLineSpacingVector(int *xspacing,
+                                         int *yspacing) const
+{
+     DFBCHECK( iface->GetLineSpacingVector (iface, xspacing, yspacing) );
+}
+
 int IDirectFBFont::GetMaxAdvance() const
 {
      int max_advance;
@@ -99,9 +105,10 @@ void IDirectFBFont::GetStringExtents (const char   *text,
 
 void IDirectFBFont::GetGlyphExtents  (unsigned int  index,
                                       DFBRectangle *rect,
-                                      int          *advance) const
+                                      int          *xadvance,
+                                      int          *yadvance) const
 {
-     DFBCHECK( iface->GetGlyphExtents (iface, index, rect, advance) );
+     DFBCHECK( iface->GetGlyphExtents (iface, index, rect, xadvance, yadvance) );
 }
 
 void IDirectFBFont::SetEncoding (DFBTextEncodingID encoding)
diff --git a/include/idirectfbfont.h b/include/idirectfbfont.h
index 9222982..250acaa 100644
--- a/include/idirectfbfont.h
+++ b/include/idirectfbfont.h
@@ -41,6 +41,8 @@ public:
      int            GetAscender      () const;
      int            GetDescender     () const;
      int            GetHeight        () const;
+     void           GetLineSpacingVector(int *xspacing,
+                                         int *yspacing) const;
      int            GetMaxAdvance    () const;
 
      void           GetKerning       (unsigned int         prev_index,
@@ -63,7 +65,8 @@ public:
                                       DFBRectangle        *ink_rect) const;
      void           GetGlyphExtents  (unsigned int         index,
                                       DFBRectangle        *rect,
-                                      int                 *advance) const;
+                                      int                 *xadvance,
+                                      int                 *yadvance) const;
 
      void           SetEncoding      (DFBTextEncodingID        encoding);
      void           EnumEncodings    (DFBTextEncodingCallback  callback,
